handshake.go

```diff
@@ -614,16 +616,18 @@ func (cfg *Config) renewDynamicCertificate(hello *tls.ClientHelloInfo, currentCe
 		return getCertWithoutReobtaining()
 	}
 
+	ctx := hello.Context()
+
 	// if the certificate hasn't expired, we can serve what we have and renew in the background
 	if timeLeft > 0 {
-		// TODO: get a proper context; we use one with timeout because retries are enabled because interactive is false
-		ctx, cancel := context.WithTimeout(context.TODO(), 5*time.Minute)
+		// we use this timeout because retries are enabled because interactive is false
+		ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
```

Comment:
🤔 I am wondering if we should keep this one set to `ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)`? `go renewAndReload(ctx, cancel)` is a background task. Do we want to cancel the task if the request gets canceled?

Reply:
This is a good question. My understanding is that this is all done within the context of the `*tls.ClientHelloInfo` - _currently_ this uses the actual `context.Context` on that struct, but that's a Go 1.17+ thing, so I'll be reworking this. 

Either way I think it still makes sense to propagate _a_ context here. It'll end up being fairly high-level/long-lived context, and we definitely want background tasks spawned here to be cancelled by it.

Reply:
I think this context is to make On-Demand TLS work (i.e. fetch a cert during the TLS handshake if the domain is eligible). Something I was mentioning to Matt recently is that it would be ideal to transfer this process to the background instead on timeout in case the issuance is slow (shouldn't be slow but it sometimes is if the ACME issuer is having problems).

I'm not saying that something different needs to be done here, but it's something to consider.

Reply:
Yeah, so this context is tricky because it is a background task even though it is initiated by a TLS handshake. It's unclear to me what the proper context should be. The duration of the handshake doesn't make sense because it's likely that the goroutine would be cancelled too early.

This one might need to stay `context.TODO()` for now until we can find the right value here.

@francislavoie 

> I think this context is to make On-Demand TLS work (i.e. fetch a cert during the TLS handshake if the domain is eligible). Something I was mentioning to Matt recently is that it would be ideal to transfer this process to the background instead on timeout in case the issuance is slow (shouldn't be slow but it sometimes is if the ACME issuer is having problems).

Not quite; this is on-demand TLS, but specifically the case when the certificate we're serving is expiring soon. We already have the cert, but it needs to be renewed (and has not yet expired). Thus we can serve it immediately and renew it in the background for performance reasons (no need to block the handshake). This already is in the background so it's OK if issuance is slow -- but that's also why we can't tether it to the current handshake's context.